### DISCUSSION TITLE
post/oop-fp/

### DISCUSSION THREAD

**Discussion by giscus** [developer-bandi/developer-bandi.github.io#8] [Announcements]
# post/oop-fp/

자주 사용되는 프로그래밍 패러다임인 OOP와 FP에 대해 살펴보고 두 패러다임을 적용한 예제를 소개합니다

https://developer-bandi.github.io/post/oop-fp/

**Comment by makeUgreat** [comment]
안녕하세요 글 읽는 중 궁금한 부분이 생겨 여쭤봅니다
'인터페이스 상속'에 대한 예시에서 `IPhone` 클래스와 `GalaxyPhone` 클래스가 `Phone` 추상 클래스를 구현하는 코드가 보이지않는데 이 경우에 셋은 완전히 독립적인 클래스인 것이 아닌가요?

**Reply by developer-bandi** [reply]
@makeUgreat 안녕하세요 질문주셔서 감사합니다. 먼저 예제 코드가 조금 명확하지 않은것같아, 조금 더 정확하게 정리해보겠습니다.

```
abstract class Phone {
  abstract call(sender:string): void;     
  abstract sendMessage(sender:string): void;
}

class IPhone extends Phone {
  call(sender:string) {
    console.log(`IPhone ${sender}에게 전화를 겁니다.`);
  }

  sendMessage(sender:string) {
    console.log(`IPhone에서 ${sender}에게 메시지를 보냅니다.`)
  }
}

class GalaxyPhone extends Phone {
  call(sender:string) {
    console.log(`GalaxyPhone에서 ${sender}에게 전화를 겁니다.`);
    this.callRecording();
  }
  sendMessage(sender:string) {
    console.log(`GalaxyPhone에서 ${sender}에게 메시지를 보냅니다.`)
  }
  private callRecording() {
    console.log('통화를 기록합니다.')
  }
  samsungpay() {
    console.log('삼성페이로 결제합니다.')
  }
}
```
위 코드에서 `IPhone`과 `GalaxyPhone`클래스는 모두 `Phone`추상클래스를 상속받고 있기에 `Phone`추상 클래스에서 명시한 `call`과 `sendMessage`메서드를 인터페이스에 맞게 구현해야합니다. 따라서 `IPhone`클래스와 `GalaxyPhone`클래스를 보면 해당 메서드가 선언되어있고 함수 본문이 채워져있으며, 인터페이스가 일치함을 알수 있습니다.(기존에 주석으로 처리된 부분도 사실 인터페이스는 일치하게 문제가 없는 코드이긴합니다.) 

따라서 현재 코드상에는 `IPhone`클래스와 `GalaxyPhone` 클래스에 `Phone` 추상 클래스를 구현하는 코드가 존재하고 `IPhone`클래스와 `GalaxyPhone` 클래스가`Phone` 추상 클래스를 상속 받고 있기 때문에 `IPhone`클래스와 `GalaxyPhone` 클래스각각은 `Phone` 추상 클래스에 의존성을 가지고 있다고 볼수 있습니다.(세 클래스가 독립적이지 않습니다.)

질문의 의도를 정확하게 이해하지 못하여 해당 내용에 대해 조금더 구체적인 설명을 적어보았는데, 만약 이해되지 않는 내용이 있다면 추가로 댓글 남겨주시면 감사하겠습니다.
